src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt

```diff
@@ -0,0 +1,554 @@
+    private var xmlCharset: Charset = Charsets.UTF_8
+    private var documentTypePublic: Boolean? = null
+    private var documentTypeFormatStringBase: String? = null
+    private var documentTypeFormatStringUrl: String? = null
+    private var xmlRootTagName: String? = null
+
+    override fun init(dictionary: IDictionaryStructure, settings: IPipelineCodecSettings?) {
+        dictionary.apply {
+            messagesTypes = messages
+            val xmlNames = HashSet<String>()
+
+            messages.forEach { (_, msgStructure) ->
+                val xmlName = msgStructure.getXmlName()
+                if (!xmlNames.add(xmlName)) {
+                    throw CodecException("Dictionary have messages with the same xml names = $xmlName")
+                }
+
+                try {
+                    checkDictionaryMessage(msgStructure)
+                } catch (e: Exception) {
+                    throw CodecException("Have wrong dictionary structure in message with name '${msgStructure.name}'", e)
+                }
+            }
+
+            getCharsetName()?.also {
+                try {
+                    xmlCharset = Charset.forName(it)
+                } catch (e: Exception) {
+                    throw CodecException("Can not find charset with name = $it")
+                }
+            }
+
+            attributes.apply {
+                documentTypePublic = get(XML_DOCUMENT_TYPE_PUBLIC_ATTRIBUTE)?.getCastValue()
+                documentTypeFormatStringBase =
+                    get(XML_DOCUMENT_TYPE_BASE_FORMAT_STR_ATTRIBUTE)?.value?.let { createFormatString(it) }
+                documentTypeFormatStringUrl =
+                    get(XML_DOCUMENT_TYPE_URL_FORMAT_STR_ATTRIBUTE)?.value?.let { createFormatString(it) }
+                xmlRootTagName = get(XML_DOCUMENT_ROOT_TAG_ATTRIBUTE)?.value
+            }
+
+            if (documentTypePublic != null && (documentTypeFormatStringUrl.isNullOrEmpty() || documentTypePublic!! && documentTypeFormatStringBase.isNullOrEmpty())) {
+                throw CodecException("Wrong dictionary. Can not set DOCTYPE tag. Base string or url string is empty")
+            }
+
+            if (documentTypePublic != null && !documentTypePublic!! && !documentTypeFormatStringBase.isNullOrEmpty()) {
+                LOGGER.warn("Using system DOCTYPE, but base string is set and will not be used")
+            }
+        }
+    }
+
+    private fun checkDictionaryMessage(msgStructure: IMessageStructure) {
+        checkDictionaryMessage(HashMap(), HashMap(), msgStructure)
+    }
+
+    private fun checkDictionaryMessage(fieldLikeNodes: MutableMap<String, Pair<String, IMessageStructure>>, fieldLikeAttrs: MutableMap<String, Pair<String, IMessageStructure>>, messageStructure: IMessageStructure) {
+        messageStructure.fields.forEach { (fieldName, fieldStructure) ->
+            if (fieldStructure.isComplex && fieldStructure is IMessageStructure && fieldStructure.isEmbedded()) {
```

Comment:
```suggestion
            if (fieldStructure.isComplex && fieldStructure.isEmbedded()) {
```
`fieldStructure is IMessageStructure` is redundant check because the structure is complex